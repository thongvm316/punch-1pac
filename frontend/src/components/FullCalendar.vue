<template>
  <div>
    <div class="calendar calendar-lg mt-4 bg-light">
      <div class="calendar-nav navbar">
        <button class="btn btn-action btn-link btn-lg" @click="lastMonth">
          <
        </button>
        <div class="navbar-primary">{{ month }} {{ year }}</div>
        <button class="btn btn-action btn-link btn-lg" @click="nextMonth">
          >
        </button>
      </div>
      <div class="calendar-container">
        <div class="calendar-header">
          <div class="calendar-date" v-for="day in days">{{ day }}</div>
        </div>
        <div class="calendar-body">
          <div class="calendar-date prev-month disabled" v-for="date in lastDaysPreviousMonth">
            <button class="date-item">{{ date }}</button>
          </div>
          <div class="calendar-date current-month" v-for="attendance in attendances">
            <button class="date-item" :class="{ 'date-today': attendance.day === today.format('YYYY-MM-DD') }">{{ attendance.day.split('-')[2] }}</button>
            <div class="calendar-events">
              <a href="#" class="calendar-event bg-primary text-light" v-if="attendance.attending_status">{{ $t(`meta.attendance_statuses.${attendance.attending_status}`) }}</a>
              <a href="#" class="calendar-event bg-error text-light" v-if="attendance.leaving_status">{{ $t(`meta.attendance_statuses.${attendance.leaving_status}`) }}</a>
              <a href="#" class="calendar-event bg-warning text-light" v-if="attendance.off_status">{{ $t(`meta.attendance_statuses.${attendance.off_status}`) }}</a>
              <a href="#" class="calendar-event bg-success text-light" v-if="attendance.holiday">{{ attendance.holiday.name }}</a>
            </div>
          </div>
          <div class="calendar-date next-month disabled" v-for="date in firstDaysNextMonth">
            <button class="date-item">{{ date }}</button>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex'

export default {
  name: 'full-calendar',

  data () {
    return {
      today: this.$moment(),
      dateContext: this.$moment(),
      days: this.$moment.weekdaysShort(),
      attendances: []
    }
  },

  computed: {
    year () {
      return this.dateContext.format('YYYY')
    },

    month () {
      return this.dateContext.format('MMMM')
    },

    initialDate () {
      return this.today.get('date')
    },

    initialMonth () {
      return this.today.format('MMMM')
    },

    initialYear () {
      return this.today.format('YYYY')
    },

    daysInMonth () {
      return this.dateContext.daysInMonth()
    },

    daysInNextMonth () {
      const previousMonth = this.$moment(this.dateContext).add(1, 'month')
      return previousMonth.daysInMonth()
    },

    daysInPreviousMonth () {
      const nextMonth = this.$moment(this.dateContext).subtract(1, 'month')
      return nextMonth.daysInMonth()
    },

    firstDayOfMonth () {
      const startDate = this.dateContext.clone().startOf('month')
      return startDate.day()
    },

    currentDate () {
      return this.dateContext.get('date')
    },

    lastDaysPreviousMonth () {
      const start = this.firstDayOfMonth
      const previousDaysInMonth = this.daysInPreviousMonth
      let previousDays = []
      for (let i = 1; i <= previousDaysInMonth; i++) {
        previousDays.push(i)
      }
      previousDays.splice(0, previousDays.length - start)
      return previousDays
    },

    firstDaysNextMonth () {
      const start = this.firstDayOfMonth
      const daysNow = this.daysInMonth
      const nextDaysInMonth = this.daysInNextMonth
      let daysNextMonth
      let nextDays = []
      for (let i = 1; i <= nextDaysInMonth; i++) {
        nextDays.push(i)
      }
      if ((start < 5 || daysNow < 30) || (start === 5 && daysNow === 30)) {
        daysNextMonth = 35 - daysNow - start
      } else {
        daysNextMonth = 42 - daysNow - start
      }
      nextDays.splice(daysNextMonth)
      return nextDays
    },

    ...mapState('initialStates', [
      'currentUser',
      'currentCompany'
    ])
  },

  methods: {
    nextMonth () {
      this.dateContext = this.$moment(this.dateContext).add(1, 'month')
      this.getCalendarAttendances(this.dateContext.format('YYYY-MM-DD')).then(response => this.formatAttendances(response))
    },

    lastMonth () {
      this.dateContext = this.$moment(this.dateContext).subtract(1, 'month')
      this.getCalendarAttendances(this.dateContext.format('YYYY-MM-DD')).then(response => this.formatAttendances(response))
    },

    formatAttendances (response) {
      this.attendances = []
      let attendances = []
      const date = response.data.attendances[0] ? this.$moment(response.data.attendances[0].day) : this.$moment()
      const userJoinDate = this.$moment(this.currentUser.created_at)
      const findHolidayByDay = function (currentDay) {
        return response.data.holidays.find(holiday => {
          return this.$moment(currentDay).isBetween(holiday.started_at, holiday.ended_at)
        })
      }

      // Loop render day in month
      for (let day = 1; day <= date.daysInMonth(); day++) {
        const currentDay = this.$moment(`${date.year()}-${date.format('MM')}-${day}`, 'YYYY-MM-D')
        let attendance = { id: null, day: currentDay.format('YYYY-MM-DD'), attended_at: '', left_at: '', attending_status: '', leaving_status: '', off_status: '', holiday: null }
        let holiday = null

        // Check if current rendering day is a holiday
        if (response.data.holidays) holiday = findHolidayByDay.call(this, currentDay)

        // If currentDay is a holiday, set holiday object into attendance object
        if (holiday) attendance = Object.assign({}, attendance, { holiday: holiday })

        // A valid day is a day before today and after current user's join date
        if (this.$moment().isAfter(currentDay) && this.$moment(currentDay).isAfter(userJoinDate)) {
          const tmpAttendance = response.data.attendances.find(item => currentDay.format('YYYY-MM-DD') === item.day)

          // If current user already attended on current day then set attendance information
          // Else if current user did not attend on current day then check if current day is weekend or unpaid leave day
          if (tmpAttendance) {
            attendance = tmpAttendance
          } else if (!holiday) {
            attendance = Object.assign({}, attendance, {
              off_status: this.currentCompany.breakdays.includes(currentDay.format('dddd').toLowerCase()) ? '' : 'unpaid_leave'
            })
          }
        }

        attendances.push(attendance)
      }

      this.attendances = attendances
    },

    ...mapActions('calendar', [
      'getCalendarAttendances'
    ])
  },

  created () {
    this.getCalendarAttendances().then(response => this.formatAttendances(response))
  }
}
</script>
