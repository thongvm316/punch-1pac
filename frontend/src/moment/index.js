import moment from 'moment-timezone'

moment.tz.setDefault(window.initialStates().company.timezone)
moment.updateLocale('en', {
  longDateFormat: {
    'L': 'MMM DD',
    'LL': 'MMM DD, YYYY',
    'LLL': 'MMM D, hh:mm',
    'LLLL': 'MMM DD YYYY, HH:mm'
  }
})

moment.updateLocale('vi', {
  monthsShort: ['Tháng 1', 'Tháng 2', 'Tháng 3', 'Tháng 4', 'Tháng 5', 'Tháng 6', 'Tháng 7', 'Tháng 8', 'Tháng 9', 'Tháng 10', 'Tháng 11', 'Tháng 12'],
  weekdaysShort: ['Chủ nhật', 'Thứ 2', 'Thứ 3', 'Thứ 4', 'Thứ 5', 'Thứ 6', 'Thứ 7'],
  longDateFormat: {
    'L': 'D MMMM',
    'LL': 'D MMMM, YYYY',
    'LLL': 'D MMMM, hh:mm',
    'LLLL': 'D MMMM YYYY, HH:mm'
  }
})
moment.locale(window.initialStates().user.language || 'en')

export default moment
