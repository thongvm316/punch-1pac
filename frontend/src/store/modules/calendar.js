import * as types from '../mutation-types'
import axios from 'axios'
import moment from '../../moment'

const state = {
  attendances: []
}

const getters = {
  getFormattedAttendances (state) {
    let attendances = []
    let index = 0
    const date = state.attendances[0] ? moment(state.attendances[0].day) : moment()

    for (let day = 1; day <= date.daysInMonth(); day++) {
      const currentDay = moment(`${date.year()}-${date.format('MM')}-${day}`, 'YYYY-MM-D')
      let attendance = null
      if (moment().isBefore(currentDay)) {
        attendance = {
          id: null,
          day: currentDay.format('YYYY-MM-DD'),
          attended_at: '',
          left_at: '',
          attending_status: '',
          leaving_status: '',
          off_status: ''
        }
      } else if (state.attendances[index] && currentDay.format('YYYY-MM-DD') === state.attendances[index].day) {
        attendance = state.attendances[index]
        index++
      } else {
        attendance = {
          id: null,
          day: currentDay.format('YYYY-MM-DD'),
          attended_at: '',
          left_at: '',
          attending_status: '',
          leaving_status: '',
          off_status: 'unpaid_leave'
        }
      }
      attendances.push(attendance)
    }

    return attendances
  }
}

const mutations = {
  [types.FETCH_CALENDAR_ATTENDANCES] (state, payload) {
    state.attendances = payload
  }
}

const actions = {
  getCalendarAttendances ({ commit }, day) {
    return axios.get('/attendances/calendar', { params: { day: day } })
                .then(response => {
                  commit(types.FETCH_CALENDAR_ATTENDANCES, response.data)
                  return response
                })
                .catch(error => { throw error })
  }
}

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions
}
