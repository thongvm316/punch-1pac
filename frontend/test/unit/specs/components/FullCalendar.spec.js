import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import initialStatesData from '../../supports/fixtures/initial-states.api'
import { fakeFebAttendances, fakeJanAttendances, fakeMarAttendances } from '../../supports/fixtures/calendar'
import moment from '@/moment'
import FullCalendar from '@/components/FullCalendar'

const nextMonth = jest.spyOn(FullCalendar.methods, 'nextMonth')
const lastMonth = jest.spyOn(FullCalendar.methods, 'lastMonth')
const currentMonth = jest.spyOn(FullCalendar.methods, 'currentMonth')
const toggleConfirmModal = jest.spyOn(FullCalendar.methods, 'toggleConfirmModal')

const getCalendarAttendances = jest.fn().mockImplementation(day => {
  if (moment(day).month() === 0) return Promise.resolve(fakeJanAttendances)
  if (moment(day).month() === 2) return Promise.resolve(fakeMarAttendances)
  else return Promise.resolve(fakeFebAttendances)
})

describe('FullCalendar.vue', () => {
  let wrapper

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when FullCalendar was mounted', () => {
    const localWrapperOps = {
      ...wrapperOps,
      data: function() {
        return {
          dateContext: moment('2019-02-05'),
          today: moment('2019-02-05')
        }
      },
      stubs: {
        CalendarDate: true,
        RequestForm: true,
        AnnualLeaveForm: true,
        PIcoPrevArrow: true,
        PIcoNextArrow: true,
        Modal: true
      },
      computed: {
        currentCompany() {
          return initialStatesData.currentCompany
        }
      },
      methods: {
        getCalendarAttendances
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(FullCalendar, localWrapperOps)
    })

    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(getCalendarAttendances).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when watch dateContext', () => {
    const localWrapperOps = {
      ...wrapperOps,
      data: function() {
        return {
          dateContext: moment('2019-01-01'),
          today: moment('2019-02-05')
        }
      },
      stubs: {
        CalendarDate: true,
        RequestForm: true,
        AnnualLeaveForm: true,
        PIcoPrevArrow: true,
        PIcoNextArrow: true,
        Modal: true
      },
      computed: {
        currentCompany() {
          return initialStatesData.currentCompany
        }
      },
      methods: {
        getCalendarAttendances
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(FullCalendar, localWrapperOps)
    })

    it('should render the other month', async () => {
      wrapper.setData({
        dateContext: moment('2019-01-01')
      })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when methods', () => {
    const localWrapperOps = {
      ...wrapperOps,
      data: function() {
        return {
          dateContext: moment('2019-02-05'),
          today: moment('2019-02-05')
        }
      },
      stubs: {
        CalendarDate: true,
        RequestForm: true,
        AnnualLeaveForm: true,
        PIcoPrevArrow: true,
        PIcoNextArrow: true,
        Modal: true
      },
      computed: {
        currentCompany() {
          return initialStatesData.currentCompany
        }
      },
      methods: {
        getCalendarAttendances
      }
    }

    beforeEach(() => {
      wrapper = shallowMount(FullCalendar, localWrapperOps)
    })

    it('should toggleConfirmModal haveBeenCalled', async () => {
      wrapper.findAll({ ref: 'calendarDate' }).at(0).trigger('click')
      await wrapper.vm.$nextTick()

      expect(toggleConfirmModal).toHaveBeenCalledWith(wrapper.vm.attendances[0])
    })

    it('should lastMonth toHaveBeenCalled', async () => {
      wrapper.find({ ref: 'lastMonthBtn' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(lastMonth).toHaveBeenCalled()
    })

    it('should nextMonth toHaveBeenCalled', async () => {
      wrapper.find({ ref: 'nextMonthBtn' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(nextMonth).toHaveBeenCalled()
    })

    it('should currentMonth haveBeenCalled', async () => {
      wrapper.find({ ref: 'currentMonthBtn' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(currentMonth).toHaveBeenCalled()
    })
  })
})
