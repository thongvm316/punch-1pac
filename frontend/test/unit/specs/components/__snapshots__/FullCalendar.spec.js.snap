// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FullCalendar.vue when FullCalendar was mounted should render correctly 1`] = `
<div>
  <div class="calendar calendar-lg">
    <div class="calendar-nav">
      <div class="calendar-nav-section">
        <h3 class="m-0">
          en
        </h3>
      </div>
      <div class="calendar-nav-section"><button class="btn btn-secondary btn-action">
          <picoprevarrow-stub></picoprevarrow-stub>
        </button> <button class="btn btn-secondary">
          Today
        </button> <button class="btn btn-secondary btn-action">
          <piconextarrow-stub></piconextarrow-stub>
        </button></div>
    </div>
    <div class="calendar-container">
      <div class="calendar-header">
        <div class="calendar-date h5 text-dark">
          Sun
        </div>
        <div class="calendar-date h5 text-dark">
          Mon
        </div>
        <div class="calendar-date h5 text-dark">
          Tue
        </div>
        <div class="calendar-date h5 text-dark">
          Wed
        </div>
        <div class="calendar-date h5 text-dark">
          Thu
        </div>
        <div class="calendar-date h5 text-dark">
          Fri
        </div>
        <div class="calendar-date h5 text-dark">
          Sat
        </div>
      </div>
      <div class="calendar-body">
        <div class="calendar-date prev-month disabled"><button class="date-item">
            27
          </button></div>
        <div class="calendar-date prev-month disabled"><button class="date-item">
            28
          </button></div>
        <div class="calendar-date prev-month disabled"><button class="date-item">
            29
          </button></div>
        <div class="calendar-date prev-month disabled"><button class="date-item">
            30
          </button></div>
        <div class="calendar-date prev-month disabled"><button class="date-item">
            31
          </button></div>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <div class="calendar-date next-month disabled"><button class="date-item">
            1
          </button></div>
        <div class="calendar-date next-month disabled"><button class="date-item">
            2
          </button></div>
      </div>
    </div>
  </div>
  <modal-stub title="What kind of request do you want to send ?">
    <div class="form-group"><label class="form-label">Kind</label> <select class="form-select">
        <option value=""></option>
        <option value="attendance">
          Request for editing attendance
        </option>
        <option value="annual_leave">
          Request a day off
        </option>}
      </select></div>
    <!---->
    <!---->
  </modal-stub>
  <modal-stub title="Add request">
    <!---->
  </modal-stub>
  <modal-stub title="Request a leave">
    <!---->
  </modal-stub>
</div>
`;

exports[`FullCalendar.vue when watch dateContext should render the other month 1`] = `
<div>
  <div class="calendar calendar-lg">
    <div class="calendar-nav">
      <div class="calendar-nav-section">
        <h3 class="m-0">
          en
        </h3>
      </div>
      <div class="calendar-nav-section"><button class="btn btn-secondary btn-action">
          <picoprevarrow-stub></picoprevarrow-stub>
        </button> <button class="btn btn-secondary">
          Today
        </button> <button class="btn btn-secondary btn-action">
          <piconextarrow-stub></piconextarrow-stub>
        </button></div>
    </div>
    <div class="calendar-container">
      <div class="calendar-header">
        <div class="calendar-date h5 text-dark">
          Sun
        </div>
        <div class="calendar-date h5 text-dark">
          Mon
        </div>
        <div class="calendar-date h5 text-dark">
          Tue
        </div>
        <div class="calendar-date h5 text-dark">
          Wed
        </div>
        <div class="calendar-date h5 text-dark">
          Thu
        </div>
        <div class="calendar-date h5 text-dark">
          Fri
        </div>
        <div class="calendar-date h5 text-dark">
          Sat
        </div>
      </div>
      <div class="calendar-body">
        <div class="calendar-date prev-month disabled"><button class="date-item">
            30
          </button></div>
        <div class="calendar-date prev-month disabled"><button class="date-item">
            31
          </button></div>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <calendardate-stub calendar-attendance="[object Object]" today="Tue Feb 05 2019 00:00:00 GMT+0700"></calendardate-stub>
        <div class="calendar-date next-month disabled"><button class="date-item">
            1
          </button></div>
        <div class="calendar-date next-month disabled"><button class="date-item">
            2
          </button></div>
      </div>
    </div>
  </div>
  <modal-stub title="What kind of request do you want to send ?">
    <div class="form-group"><label class="form-label">Kind</label> <select class="form-select">
        <option value=""></option>
        <option value="attendance">
          Request for editing attendance
        </option>
        <option value="annual_leave">
          Request a day off
        </option>}
      </select></div>
    <!---->
    <!---->
  </modal-stub>
  <modal-stub title="Add request">
    <!---->
  </modal-stub>
  <modal-stub title="Request a leave">
    <!---->
  </modal-stub>
</div>
`;
